cpp/cmake_modules/DefineOptions.cmake

```diff
@@ -110,6 +110,12 @@ macro(resolve_option_dependencies)
   if(MSVC_TOOLCHAIN)
     set(ARROW_USE_GLOG OFF)
   endif()
+  # Tests are crashed with mold + sanitizer checks.
+  if(ARROW_USE_ASAN
+     OR ARROW_USE_TSAN
+     OR ARROW_USE_UBSAN)
+    set(ARROW_USE_MOLD OFF)
```

Comment:
Should this not produce a message in case the user switched it on?

Reply:
OK. I've added a warning message for this case.